### PR TITLE
Update wallet coin selection integration to latest rust-coinselect

### PR DESCRIPTION
**Pull request by NeoZ666**
# Pull Request
Update wallet coin selection integration to latest `rust-coinselect`

## Description

This PR updates Coinswap’s coin selection flow to align with the current `rust-coinselect` API and behavior.

**What changed**
- Updates wallet coin selection to use ranked `select_coin` results and the new error model
- Passes destination/output address type into `coin_select` so output weight estimation matches P2WPKH vs P2TR spends
- Refactors manual UTXO handling to account for actual selected input weight and fee impact
- Removes outdated split-based and fallback selection logic that is no longer part of the protocol
- Improves insufficient-funds propagation by using coinselect’s structured errors directly

## Related Issue(s)
Closes #

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [ ] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [ ] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [ ] Both
- [ ] Neither (infrastructure / tooling only)

> **Note:** When modifying protocol logic, changes must be made to **both** versions unless the
> scope is explicitly version-specific.

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [ ] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [ ] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [ ] I ran `cargo +nightly fmt --all` and committed the result
- [ ] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [ ] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [ ] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [ ] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [ ] All unit tests pass (`cargo test`)
- [ ] Integration tests pass (`cargo test --features integration-test`)
- [ ] Changes were manually tested on **regtest**
- [ ] End-to-end maker ↔ taker swap tested (where applicable)
- [ ] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [ ] Complex logic is commented

### Security & Privacy (Critical)
- [ ] This change preserves **trustlessness** and **atomic swap guarantees**
- [ ] No regression in **sybil resistance** or **fidelity bond** logic
- [ ] Tor anonymity and P2P message flow were reviewed
- [ ] No new attack vectors or trust assumptions introduced
- [ ] Edge cases and error handling considered
- [ ] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [ ] `integration-test` feature flag is not reachable in production code paths

## How to Test

```bash
# Standard swap (good baseline)
cargo test --test standard_swap --features integration-test -- --nocapture

# Abort scenarios
cargo test --test abort1 --features integration-test -- --nocapture
cargo test --test taproot_taker_abort1 --features integration-test -- --nocapture

# Malicious behavior & recovery
cargo test --test malice1 --features integration-test -- --nocapture
cargo test --test taproot_timelock_recovery --features integration-test -- --nocapture

# Or run the full suite
cargo test --features integration-test
```

## Summary by CodeRabbit

* **New Features**
  * Wallet coin selection now infers and applies the destination address type (e.g., P2WPKH vs P2TR) across send, funding, fidelity, and split-UTXO flows.
  * Coin selection is now more weight/VB aware for improved fee accuracy.

* **Bug Fixes**
  * Manual coin selection is stricter, rejecting mixed UTXO types and unavailable/invalid outpoints.
  * Insufficient-funds reporting/selection now prioritizes the most relevant shortage scenario.

* **Tests**
  * Updated integration tests, assertions, and the wallet basic example to match the improved selection behavior.

### CODE REVIEW

File: src/wallet/api.rs [hunk 1]
```diff
@@ -260,6 +260,14 @@ impl Display for UTXOSpendInfo {
     }
 }
 
+pub(crate) fn infer_address_type(script_pubkey: &Script) -> AddressType {
+    if script_pubkey.is_p2tr() {
+        AddressType::P2TR
+    } else {
+        AddressType::P2WPKH
+    }
+}
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply src/wallet/api.rs:266]: Future work.

File: src/wallet/api.rs [hunk 2]
```diff
@@ -2300,38 +2219,32 @@ impl Wallet {
...
-                base_weight: tx_base_weight,
+                min_absolute_fee: 0,
+                base_weight,
                 change_weight: change_weight.to_wu(),
                 change_cost: cost_of_change,
-                avg_input_weight,
-                avg_output_weight,
                 min_change_value: 294, // Minimal NonDust value: 294
                 excess_strategy: ExcessStrategy::ToChange,
             };
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply src/wallet/api.rs:2219]: Done

File: tests/integration/coinselect_no_change.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+        println!(
+            "PASS  [{description}]  outputs={}  payment={payment_sats}",
+            tx.output.len()
+        );
+    }
+
+    test_framework.stop();
+    block_generation_handle.join().unwrap();
+}
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply tests/integration/coinselect_no_change.rs]: A transcient test. Since my machine is occupied with testing for other pr. Removed it already since everything works.

File: tests/integration/coinselect_no_change.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                true,
+                Amount::from_sat(payment_sats),
+                vec![dest_addr],
+                Amount::from_sat(2),
+                None,
+                None,
+            )
+            .unwrap_or_else(|e| {
+                panic!("[{description}] create_funding_txes_regular_swaps failed: {e:?}", description=description, e=e)
+            });
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply tests/integration/coinselect_no_change.rs]: Doesn't output anything

File: src/wallet/api.rs [hunk 5]
```diff
@@ -2300,38 +2219,32 @@ impl Wallet {
...
+            let base_weight = tx_weight_with_selected_inputs(selected_weight);
 
-            let remaining_target = (amount.to_sat() + estimated_fee).saturating_sub(selected_total);
+            let remaining_target = target.saturating_sub(selected_total).max(1);
 
-            // Create coin selection options with adjusted target
             let coin_selection_option = CoinSelectionOpt {
                 target_value: remaining_target,
-                target_feerate: feerate as f32 / 4.0, //sats per wu
+                target_feerate: target_feerate_wu, //sats per wu
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply src/wallet/api.rs:2211]: incorrect. Its ==0. Not < 0

File: src/wallet/fidelity.rs [hunk 6]
```diff
@@ -483,7 +481,13 @@ impl Wallet {
...
         let (index, fidelity_addr, fidelity_pubkey) = self.get_next_fidelity_address(locktime)?;
 
-        let coins = self.coin_select(amount, feerate, None, None)?;
+        let coins = self.coin_select(
+            amount,
+            feerate,
+            infer_address_type(&fidelity_addr.script_pubkey()),
+            None,
+            None,
+        )?;
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply src/wallet/fidelity.rs:492]: Done

File: src/wallet/api.rs [hunk 7]
```diff
@@ -1948,70 +1953,25 @@ impl Wallet {
         &self,
         amount: Amount,
         feerate: f64,
+        output_address_type: AddressType,
         manually_selected_outpoints: Option<Vec<OutPoint>>,
         excluded_outpoints: Option<Vec<OutPoint>>,
     ) -> Result<Vec<(ListUnspentResultEntry, UTXOSpendInfo)>, WalletError> {
-        // P2TR input weight breakdown:
```
Comment (NeoZ666) [inline src/wallet/api.rs:1969]: This data for inputs is now being pulled directly from estimate_witness_data in the same mod

File: src/wallet/api.rs [hunk 8]
```diff
@@ -260,6 +256,15 @@ impl Display for UTXOSpendInfo {
     }
 }
 
+pub(crate) fn infer_address_type(script_pubkey: &Script) -> AddressType {
```
Comment (NeoZ666) [inline src/wallet/api.rs:259]: This is a helper function for inferring the address type of outputs to coinselect. Change is currently P2TR, but we can either 
a. Infer it's type via the payment output.
b. Infer it's type via the input set (this will mitigate the problem of script mismatch)
c. Infer it's type via wallet state.

File: tests/integration/funding_dynamic_splits.rs [hunk 9]
```diff
@@ -21,20 +21,20 @@ const UTXO_SETS: &[&[u64]] = &[
...
-    (708_742, &[9091, 38012, 53245, 91379, 107831, 109831, 301909, 712971], 4),
-    (500_000, &[3919, 9091, 35892, 38012, 46824, 53245, 65658, 100000, 107831, 109831, 432441], 4),
-    (654_321, &[3919, 38012, 46824, 53245, 65658, 70000, 91379, 100000, 107831, 301909, 432441], 4),
-    (90_000, &[3919, 35892, 53245, 91379], 4),
-    (10_000, &[3919, 9091], 4),
-    (1_000_000, &[9091, 35892, 65658, 70000, 91379, 100000, 298092, 432441, 900000], 4),
-    (999_999, &[9091, 35892, 65658, 70000, 91379, 100000, 298092, 432441, 900000], 4),
-    (123_456, &[9091, 35892, 38012, 46824, 53245, 70000], 4),
-    (250_000, &[3919, 35892, 46824, 53245, 65658, 91379, 100000, 107831], 4),
+    (54_082, &[46824, 65658], 4),
```
Comment (NeoZ666) [inline tests/integration/funding_dynamic_splits.rs:24]: This is due to keeping minimizing weight as the best policy. However, waste priority would provide a balance between privacy and provide a long term longevity of wallet, given it's informed by statistical data. Later on, we use `change_cost`, `long_term_feerate` as a penalizing force to shape the input and output sets towards heavy density utxo clusters on the network.

File: src/wallet/api.rs [hunk 10]
```diff
@@ -2300,38 +2189,32 @@ impl Wallet {
...
+                min_change_value: 330, // P2TR dust threshold (since P2WPKH's 294)
                 excess_strategy: ExcessStrategy::ToChange,
             };
 
-            // Run coin selection on single addresses only
             match select_coin(&single_output_groups, &coin_selection_option) {
-                Ok(selection) => {
-                    let additional_utxos: Vec<_> = selection
+                Ok(results) => {
+                    let (_, result) = results.into_iter().next().unwrap();
```
Comment (NeoZ666) [inline src/wallet/api.rs:2223]: Coinselect now outputs the result of all the algorithms which would be important to derive the rules for a top level selection policy.

File: src/wallet/api.rs [hunk 11]
```diff
@@ -2042,129 +2001,101 @@ impl Wallet {
...
         };
 
-        let target_weight = Weight::from_wu(TARGET_OUTPUT_WEIGHT);
-        let avg_output_weight = (change_weight.to_wu() + target_weight.to_wu()) / 2;
-
-        // Try regular UTXOs first, then fall back to swap UTXOs if selection fails
-        let utxo_types_to_try = if manual_regular_selected {
+        let tx_weight_with_selected_inputs = |selected_weight: u64| {
+            BASE_TXN_ONLY_WEIGHT + selected_weight + target_output_weight + change_output_weight
+        };
```
[the remark this answers is not in the document]
Comment (NeoZ666) [reply src/wallet/api.rs]: done

File: src/wallet/api.rs [hunk 12]
```diff
@@ -1963,70 +1953,25 @@ impl Wallet {
...
-        // - Num witness elements (1 byte)    = 1 WU
-        // - Schnorr signature (64 bytes)     = 64 WU
-        // Subtotal witness:                  = 65 WU
-
-        // Total: 164 + 65 = 229 WU
-        // Adding 2 bytes as a buffer : 229 + 2 = 231 WU
-        const P2TR_INPUT_WEIGHT: u64 = 231; // Total weight units
-
-        // P2TR script-pubkey size:
-        // - OP_1 (1 byte)
```
Comment (mojoX911) [inline src/wallet/api.rs:1988]: These comments were important. Bring them back and adjust them to match with your new values.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Some ornamental changes needed.

File: src/wallet/api.rs (hunk 12 above)
Comment (NeoZ666) [reply src/wallet/api.rs:1988]: Sure

File: src/wallet/api.rs [hunk 13]
```diff
@@ -2057,129 +2001,100 @@ impl Wallet {
...
-                        .any(|manual_utxo| {
-                            OutPoint::new(manual_utxo.txid, manual_utxo.vout) == outpoint
-                        })
+                    manual_outpoints.contains(&outpoint)
                 });
 
-            let unspents = non_manual_unspents.into_iter().cloned().collect::<Vec<_>>();
+            let unspents = candidate_utxos.into_iter().cloned().collect::<Vec<_>>();
 
-            // Group UTXOs by address
```
Comment (mojoX911) [inline src/wallet/api.rs:2182]: bring back these sectional comments to clearly define which section of the new algo is doing what.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
